libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.h

```diff
@@ -0,0 +1,284 @@
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#import <Foundation/Foundation.h>
+#import "SFOAuthCoordinator.h"
+#import "SFSDKOAuthClientContext.h"
+
+@class SFSDKOAuthClientContext;
+@class SFSDKMutableOAuthClientContext;
+@class SFSDKOAuthClient;
+@class WKWebView;
+@class SFIdentityCoordinator;
+@class SFOAuthCredentials;
+@class SFOAuthInfo;
+@class SFLoginViewController;
+@class SFSDKOAuthViewHandler;
+@class SFSDKWindowContainer;
+@class SFSDKOAuthClientAdvanced;
+@class SFSDKOAuthClientIDP;
+@class SFAuthErrorHandlerList;
+@class SFAuthErrorHandler;
+@class SFSDKOAuthClientConfig;
+@class SFIdentityData;
+NS_ASSUME_NONNULL_BEGIN
+
+/** Delegate that will be used to notify of all the OAuth Client events.
+ */
+@protocol SFSDKOAuthClientDelegate <NSObject>
+
+@optional
+/**
+ * Called when client will begin authentication.
+ * @param client  initiating the request
+ */
+- (void)authClientWillBeginAuthentication:(SFSDKOAuthClient *)client;
+
+/**
+ * Called when client will begin authentication.
+ * @param client making the call prior to refresh credentials
+ */
+- (void)authClientWillRefreshCredentials:(SFSDKOAuthClient *)client;
+
+/**
+ * Called when client will begin authentication.
+ * @param client making the call prior to revoke credentials
+ */
+- (void)authClientWillRevokeCredentials:(SFSDKOAuthClient *)client;
+
+/**
+ * Called when client will begin authentication.
+ * @param client making the call prior after the  revoke of credentials
+ */
+- (void)authClientDidRevokeCredentials:(SFSDKOAuthClient *)client;
+
+/**
+ * Called when client will begin authentication.
+ * @param client making the call
+ * @param error describes the error that occurred
+ */
+- (void)authClientDidFail:(SFSDKOAuthClient *)client error:(NSError *_Nullable)error;
+
+/**
+ * Called by the underlying mechanism allow for detection of network state
+ * @param client making the call
+ */
+- (BOOL)authClientIsNetworkAvailable:(SFSDKOAuthClient *)client;
+
+/**
+ * Called when client did finish authentication.
+ * @param client making the call
+ */
+- (void)authClientDidFinish:(SFSDKOAuthClient *)client;
+
+/**
+ Called when error occurs offline.
+ @param client The instance of SFSDKOAuthClient making the call.
+ */
+- (void)authClientContinueOfflineMode:(SFSDKOAuthClient *)client;
+
+/**
+ Called when a generic flow authentication is cancelled.
+ @param client The instance of SFSDKOAuthClient making the call.
+ */
+- (void)authClientDidChangeLoginHost:(SFSDKOAuthClient *)client loginHost:(NSString *)newLoginHost;
```

Comment:
Do we need this? This is probably handled in the notification mechanism of the persistence layer.

Reply:
@bhariharan This is required for the IDP client to notify its delegate of change of host that occurred. Should result in a re-login flow. The client doesn't directly make the change to the persistent store.